src/OpenTelemetry.Extensions.PersistentStorage.Abstractions/PersistentBlob.cs

```diff
@@ -34,6 +34,7 @@ public abstract class PersistentBlob
         /// </returns>
         public bool TryRead(out byte[] buffer)
```

Comment:
A couple things with the null reference annotations...

* You won't get good nullable analysis in this project because it is only targeting netstandard2.0 & net462. You can add a netstandard2.1 target, something >= netcoreapp2.1, or do something like this: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/584654cb967bc8f6a65d9e9055f486ab7c2efe93/src/OpenTelemetry.Extensions/OpenTelemetry.Extensions.csproj#L5

* I don't think the "Try" methods are correctly annotated. Here is what I would do:

```csharp
public abstract class PersistentBlobProvider
{
   public bool TryCreateBlob(byte[] buffer, int leasePeriodMilliseconds, [NotNullWhen(true)] out PersistentBlob? blob)
   public bool TryCreateBlob(byte[] buffer, [NotNullWhen(true)] out PersistentBlob? blob)
   public bool TryGetBlob([NotNullWhen(true)] out PersistentBlob? blob)

   protected abstract bool OnTryCreateBlob(byte[] buffer, int leasePeriodMilliseconds, [NotNullWhen(true)] out PersistentBlob? blob);
   protected abstract bool OnTryCreateBlob(byte[] buffer, [NotNullWhen(true)] out PersistentBlob? blob);
   protected abstract bool OnTryGetBlob([NotNullWhen(true)] out PersistentBlob? blob);
}

public abstract class PersistentBlob
{
   public bool TryRead([NotNullWhen(true)] out byte[]? buffer)

   protected abstract bool OnTryRead([NotNullWhen(true)] out byte[]? buffer);
}
```

The idea is when the "Try" fails, then the "out" thing will be `null`. If the "Try" succeeds, the "out" thing will NOT be `null`. `NotNullWhenAttribute` is a special attribute to help the analysis engine understand the behavior.

Where it gets fun is `NotNullWhenAttribute` only exists on newer targets. So you either have to define it internal for the targets where it doesn't exist, or wrap all of those in preprocessor like...

```csharp
public bool TryGetBlob(
#if NETSTANDARD2_1 || NETCOREAPP3_1_OR_GREATER
   [NotNullWhen(true)]
#endif
   out PersistentBlob? blob)
```

Reply:
@CodeBlanch - Updated - Please review if I understood your suggestion correctly. I am not clear on 1.